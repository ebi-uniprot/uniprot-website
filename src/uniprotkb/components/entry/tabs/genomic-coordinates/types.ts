export type GenomicEntry = {
  accession: string;
  name: string;
  taxid: number; // Swagger says optional, but apparently not
  sequence: string;
  /**
   * @deprecated shouldn't use this from this endpoint
   */
  protein: any; // eslint-disable-line @typescript-eslint/no-explicit-any
  /**
   * @deprecated shouldn't use this from this endpoint
   */
  gene?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
  gnCoordinate?: Array<GenomicCoordinate>;
};

export type GenomicCoordinate = {
  genomicLocation: GenomicLocation;
  /**
   * @deprecated shouldn't use this from this endpoint
   */
  feature?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
  // Sometimes not from Ensembl
  ensemblGeneId?: string;
  // Sometimes not from Ensembl
  ensemblTranscriptId?: string;
  // Sometimes not from Ensembl
  ensemblTranslationId?: string;
};

export type GenomicLocation = {
  exon: Array<ExonMap>;
  chromosome?: string;
  start: number; // Swagger says optional, but apparently not
  end: number; // Swagger says optional, but apparently not
  reverseStrand?: boolean;
};

type ExonMap = {
  proteinLocation: Location; // Swagger says optional, but apparently not
  genomeLocation: Location; // Swagger says optional, but apparently not
  id?: string;
};

type Location = {
  begin: Position; // Swagger says optional, but apparently not
  end: Position; // Swagger says optional, but apparently not
  position?: Position; // ?
  sequence?: string;
};

type Position = {
  position: number; // Swagger says optional, but apparently not
  status?: string;
  evidence?: number[];
};
