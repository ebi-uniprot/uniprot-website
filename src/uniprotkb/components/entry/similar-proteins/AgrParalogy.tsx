import cn from 'classnames';
import { ExternalLink, Loader } from 'franklin-sites';
import { Link } from 'react-router-dom';

import { Location, LocationToPath } from '../../../../app/config/urls';
import ErrorHandler from '../../../../shared/components/error-pages/ErrorHandler';
import TableFromData, {
  TableFromDataColumn,
} from '../../../../shared/components/table/TableFromData';
import WithTooltip from '../../../../shared/components/WithTooltip';
import externalUrls from '../../../../shared/config/externalUrls';
import useDataApi from '../../../../shared/hooks/useDataApi';
import * as logging from '../../../../shared/utils/logging';
import { stringifyQuery } from '../../../../shared/utils/url';
import {
  AgrParalogs,
  AgrParalogsResult,
  PredictionMethodsMatchedName,
} from '../../../types/agrParalogs';
import styles from './styles/agr-homology.module.scss';

const getTaxonQuery = (curie: string): string | null => {
  const reTaxonId = /NCBITaxon:(?<taxonId>\d+)/i;
  const match = curie.match(reTaxonId);
  if (!match?.groups?.taxonId) {
    return null;
  }
  return `(taxonomy_id:${match?.groups?.taxonId})`;
};

/*
MGI:88059               (xref:mgi-88059)
Xenbase:XB-GENE-479154  (xref:xenbase-XB-GENE-479154)
FB:FBgn0000108          (xref:flybase-FBgn0000108) 
RGD:2139                (xref:rgd-2139)
ZFIN:ZDB-GENE-000616-13 (xref:zfin-ZDB-GENE-000616-13)
HGNC:620                (xref:hgnc-620)
WB:WBGene00000149       (xref:agr-WBGene00000149) 
*/
const xrefTokenToQueryPrefix = new Map([
  ['MGI', 'mgi'],
  ['Xenbase', 'xenbase'],
  ['FB', 'flybase'],
  ['RGD', 'rgd'],
  ['ZFIN', 'zfin'],
  ['HGNC', 'hgnc'],
  ['WB', 'agr'],
]);

const getXrefQuery = (primaryExternalId: string) => {
  const [xrefToken, ...idToken] = primaryExternalId.split(':');
  if (!idToken) {
    logging.error(
      `No token found for AGR primaryExternalId: ${primaryExternalId}`
    );
    return null;
  }
  const queryPrefix = xrefTokenToQueryPrefix.get(xrefToken);
  if (!queryPrefix) {
    logging.error(
      `No query prefix found for AGR primaryExternalId: ${primaryExternalId}`
    );
    return null;
  }
  return `(xref:${queryPrefix}-${idToken})`;
};

const getXrefAndTaxonQuery = (
  gene: AgrParalogsResult['geneToGeneParalogy']['objectGene']
) => {
  const taxonQuery = getTaxonQuery(gene.taxon.curie);
  const xrefQuery = getXrefQuery(gene.primaryExternalId);
  return taxonQuery && xrefQuery ? `${taxonQuery} AND ${xrefQuery}` : null;
};

// Lifted from https://github.com/alliance-genome/agr_ui/blob/6f5acc104df6274bb0642a2317a5b6b102a91b32/src/components/homology/constants.js#L1

const PARALOGY_METHODS: {
  method: PredictionMethodsMatchedName;
  tooltip: string;
}[] = [
  {
    method: 'Ensembl Compara',
    tooltip: 'Paralogy relationships inferred by the Ensembl Compara pipeline.',
  },
  {
    method: 'HGNC',
    tooltip:
      'Paralogs manually curated by the HUGO Gene Nomenclature Committee for human gene family relationships.',
  },
  {
    method: 'InParanoid',
    tooltip:
      'In-paralogs identified within species using the InParanoid clustering algorithm.',
  },
  {
    method: 'OMA',
    tooltip:
      'Paralogy relationships predicted by the OMA (Orthologous MAtrix) algorithm through internal duplication detection.',
  },
  {
    method: 'OrthoFinder',
    tooltip:
      'Paralog assignments determined by OrthoFinder’s graph-based gene-family inference.',
  },
  {
    method: 'OrthoInspector',
    tooltip:
      'Paralogs detected using the OrthoInspector homology-detection framework.',
  },
  {
    method: 'PANTHER',
    tooltip:
      'Paralog relationships assigned by PANTHER’s phylogenetic family and subfamily models.',
  },
  {
    method: 'PhylomeDB',
    tooltip:
      'Paralogy calls supported by the PhylomeDB gene-tree database through duplication nodes.',
  },
  {
    method: 'SonicParanoid',
    tooltip:
      'Paralogs inferred by the SonicParanoid pipeline leveraging fast pairwise comparisons.',
  },
  {
    method: 'SGD',
    tooltip:
      'Paralogous gene relationships manually curated by the Saccharomyces Genome Database for S. cerevisiae.',
  },
];

const columns: TableFromDataColumn<AgrParalogsResult>[] = [
  {
    id: 'gene-symbol',
    label: (
      <WithTooltip tooltip="Gene symbol of the ortholog in the target species.">
        Gene Symbol
      </WithTooltip>
    ),
    render: (data) => {
      const query = getXrefAndTaxonQuery(data.geneToGeneParalogy.objectGene);
      const gene = data.geneToGeneParalogy.objectGene.geneSymbol.displayText;

      return query ? (
        <Link
          to={{
            pathname: LocationToPath[Location.UniProtKBResults],
            search: stringifyQuery({
              query,
            }),
          }}
        >
          {gene}
        </Link>
      ) : (
        gene
      );
    },
  },
  {
    id: 'rank',
    label: (
      <WithTooltip tooltip="Gene symbol of the ortholog in the target species.">
        Rank
      </WithTooltip>
    ),
    render: (data) => data.geneToGeneParalogy.rank,
  },
  {
    id: 'length',
    label: (
      <WithTooltip tooltip="Gene symbol of the ortholog in the target species.">
        Length
      </WithTooltip>
    ),
    render: (data) => data.geneToGeneParalogy.length,
  },
  {
    id: 'similarity',
    label: (
      <WithTooltip tooltip="Gene symbol of the ortholog in the target species.">
        Similarity
      </WithTooltip>
    ),
    render: (data) => data.geneToGeneParalogy.similarity,
  },
  {
    id: 'identity',
    label: (
      <WithTooltip tooltip="Gene symbol of the ortholog in the target species.">
        Identity
      </WithTooltip>
    ),
    render: (data) => data.geneToGeneParalogy.identity,
  },
];
for (const [index, { method, tooltip }] of PARALOGY_METHODS.entries()) {
  columns.push({
    id: method,
    label:
      index === 0 ? (
        <div className={styles['methods-label-container']}>
          <span className={styles['methods-label']}>
            <WithTooltip tooltip="Result of paralogy-inference resource and algorithm methods.">
              Method
            </WithTooltip>
          </span>
          <div className={styles['method-label']}>
            <WithTooltip tooltip={tooltip}>{method}</WithTooltip>
          </div>
        </div>
      ) : (
        <div className={styles['method-label']}>
          <WithTooltip tooltip={tooltip}>{method}</WithTooltip>
        </div>
      ),
    render: (data) => {
      const predictionMethodsMatchedSet = new Set(
        data.geneToGeneParalogy.predictionMethodsMatched?.map((m) => m.name)
      );
      const predictionMethodsNotMatchedSet = new Set(
        data.geneToGeneParalogy.predictionMethodsNotMatched?.map((m) => m.name)
      );
      let symbol: string, title: string;
      if (predictionMethodsMatchedSet.has(method)) {
        symbol = '●';
        title = `Match by ${method}`;
      } else if (predictionMethodsNotMatchedSet.has(method)) {
        symbol = '○';
        title = `No match by ${method}`;
      } else {
        symbol = '-';
        title = `Comparison not available for ${method}`;
      }
      return (
        <span key={method} title={title} className={styles['method-render']}>
          {symbol}
        </span>
      );
    },
  });
}
columns.push({
  id: 'method-match-count',
  label: (
    <span className={styles['method-match-count-label']}>
      <WithTooltip tooltip="Number of independent paralogy-inference resource and algorithm methods that support this gene pair.">
        Match count
      </WithTooltip>
    </span>
  ),
  render: (data) => {
    const scoreNumerator =
      data.geneToGeneParalogy.predictionMethodsMatched.length;
    const scoreDenominator =
      scoreNumerator +
      (data.geneToGeneParalogy.predictionMethodsNotMatched?.length || 0);
    return (
      <span
        title={`${scoreNumerator} matches from ${scoreDenominator} checked methods (${Math.round((100 * scoreNumerator) / scoreDenominator)}%)`}
        className={styles['method-match-count-render']}
      >
        {scoreNumerator} of {scoreDenominator}
      </span>
    );
  },
});

const getRowId = (data: AgrParalogsResult) =>
  data.geneToGeneParalogy.objectGene.primaryExternalId;

type Props = {
  agrId: string;
};

const AgrParalogy = ({ agrId }: Props) => {
  const { data, loading, error, status, progress } = useDataApi<AgrParalogs>(
    agrId ? externalUrls.AgrHomologs(agrId, 'paralogs') : null
  );

  if (loading) {
    return <Loader progress={progress} />;
  }

  if (error) {
    return <ErrorHandler status={status} error={error} />;
  }

  if (!data?.results?.length) {
    return 'No Orthology data is available from the Alliance of Genome Resources.';
  }

  return (
    <>
      <div className={styles['agr-link']}>
        The data in this table is sourced from the Alliance of Genome Resources.
        <br />
        <ExternalLink url={externalUrls.AgrEntryHomologs(agrId, 'paralogy')}>
          View source data
        </ExternalLink>
      </div>
      <TableFromData
        id="agr-orthology"
        columns={columns}
        data={data.results}
        getRowId={getRowId}
        className={cn(
          styles['agr-homology-table'],
          styles['agr-paralogy-table']
        )}
      />
    </>
  );
};

export default AgrParalogy;
