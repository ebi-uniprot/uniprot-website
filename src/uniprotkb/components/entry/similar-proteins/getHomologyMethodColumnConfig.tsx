import AgrHomologyMatch from './AgrHomologyMatch';
import styles from './styles/agr-homology.module.scss';

const AgrMethodLabel: React.FC<{
  method: React.ReactNode;
}> = ({ method }) => (
  <div className={styles['method-label']}>
    <span>{method}</span>
  </div>
);

type PredictionMethods = {
  internal: boolean;
  obsolete: boolean;
  name: string;
  notInternalOrObsolete: boolean;
};

type HomologyGetter<T> = (data: T) => {
  predictionMethodsMatched: PredictionMethods[];
  predictionMethodsNotMatched?: PredictionMethods[];
};

function getHomologyMethodColumnConfig<T>(
  index: number,
  method: string,
  homologyGetter: HomologyGetter<T>
) {
  return {
    id: method,
    label:
      index === 0 ? (
        <div className={styles['methods-label-container']}>
          <span className={styles['methods-label']}>Method</span>
          <AgrMethodLabel method={method} />
        </div>
      ) : (
        <AgrMethodLabel method={method} />
      ),
    render: (data: T) => (
      <AgrHomologyMatch
        method={method}
        matched={homologyGetter(data).predictionMethodsMatched}
        notMatched={homologyGetter(data).predictionMethodsNotMatched}
      />
    ),
  };
}

export default getHomologyMethodColumnConfig;
