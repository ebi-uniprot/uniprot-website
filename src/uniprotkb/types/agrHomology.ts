export type PredictionMethods<T extends string> = {
  internal: boolean;
  obsolete: boolean;
  name?: T;
  notInternalOrObsolete: boolean;
  displayText?: string;
};
