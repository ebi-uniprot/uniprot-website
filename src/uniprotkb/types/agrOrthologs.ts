export type AgrOrthologs = {
  results: AgrOrthologsResult[];
  total: number;
  returnedRecords: number;
  requestDate: string;
};

export type AgrOrthologsResult = {
  category: Category;
  searchable: boolean;
  stringencyFilter: StringencyFilter;
  geneAnnotations: GeneAnnotation[];
  geneAnnotationsMap: { [key: string]: GeneAnnotation };
  geneToGeneOrthologyGenerated: GeneToGeneOrthologyGenerated;
};

type Category = 'gene_to_gene_orthology';

type GeneAnnotation = {
  hasDiseaseAnnotations: boolean;
  hasExpressionAnnotations: boolean;
  geneIdentifier: string;
};

type GeneToGeneOrthologyGenerated = {
  internal: boolean;
  obsolete: boolean;
  subjectGene: Gene;
  objectGene: Gene;
  isBestScore: Score;
  isBestScoreReverse: Score;
  confidence: Confidence;
  strictFilter: boolean;
  moderateFilter: boolean;
  predictionMethodsMatched: Method[];
  predictionMethodsNotMatched?: Method[];
  predictionMethodsNotCalled?: Method[];
  notInternalOrObsolete: boolean;
};

type Confidence = {
  internal: boolean;
  obsolete: boolean;
  name?: 'moderate' | 'low' | 'high';
  notInternalOrObsolete: boolean;
  displayText?: string;
};

type Score = {
  internal: false;
  obsolete: false;
  name: 'Yes' | 'No';
  notInternalOrObsolete: true;
};

export type PredictionMethodName =
  | 'Hieranoid'
  | 'SonicParanoid'
  | 'InParanoid'
  | 'Xenbase'
  | 'ZFIN'
  | 'HGNC'
  | 'OrthoFinder'
  | 'PANTHER'
  | 'OMA'
  | 'Ensembl Compara'
  | 'OrthoInspector'
  | 'PhylomeDB';

type Gene = {
  type: Type;
  internal: boolean;
  obsolete: boolean;
  primaryExternalId: string;
  taxon: Taxon;
  geneSymbol: GeneSymbol;
  notInternalOrObsolete: boolean;
};

type Method = {
  internal: boolean;
  obsolete: boolean;
  name: PredictionMethodName;
  notInternalOrObsolete: boolean;
};

type GeneSymbol = {
  internal: boolean;
  obsolete: boolean;
  displayText: string;
  notInternalOrObsolete: boolean;
};

type Taxon = {
  internal: boolean;
  obsolete: boolean;
  curie: string;
  name: string;
  childCount: number;
  descendantCount: number;
  notInternalOrObsolete: boolean;
};

type Type = 'Gene';

type StringencyFilter = 'all' | 'stringent';
