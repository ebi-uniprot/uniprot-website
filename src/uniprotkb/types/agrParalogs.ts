export type AgrParalogs = {
  results: AgrParalogsResult[];
  total: number;
  returnedRecords: number;
  requestDuration: string;
  requestDate: string;
};

export type AgrParalogsResult = {
  category: 'gene_to_gene_paralogy';
  searchable: boolean;
  geneToGeneParalogy: GeneToGeneParalogy;
};

type GeneToGeneParalogy = {
  internal: boolean;
  obsolete: boolean;
  subjectGene: Gene;
  objectGene: Gene;
  predictionMethodsMatched: PredictionMethod[];
  predictionMethodsNotMatched: PredictionMethod[];
  predictionMethodsNotCalled: PredictionMethod[];
  rank: number;
  length: number;
  similarity: number;
  identity: number;
  notInternalOrObsolete: boolean;
};

type Gene = {
  type: 'Gene';
  internal: boolean;
  obsolete: boolean;
  primaryExternalId: string;
  taxon: Taxon;
  geneSymbol: GeneSymbolName;
  geneFullName: GeneSymbolName;
  notInternalOrObsolete: boolean;
  dateCreated: string;
};

type GeneSymbolName = {
  internal: boolean;
  obsolete: boolean;
  formatText: string;
  displayText: string;
  notInternalOrObsolete: boolean;
};

type Taxon = {
  internal: boolean;
  obsolete: boolean;
  curie: string;
  name: string;
  childCount: number;
  descendantCount: number;
  notInternalOrObsolete: boolean;
};

export type PredictionMethod = {
  internal: boolean;
  obsolete: boolean;
  name: PredictionMethodName;
  notInternalOrObsolete: boolean;
};

export type PredictionMethodName =
  | 'Ensembl Compara'
  | 'OrthoFinder'
  | 'InParanoid'
  | 'PhylomeDB'
  | 'PANTHER'
  | 'HGNC'
  | 'SGD'
  | 'OrthoInspector'
  | 'OMA'
  | 'SonicParanoid';
