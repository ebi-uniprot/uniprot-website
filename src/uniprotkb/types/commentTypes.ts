import { Xref } from '../../shared/types/apiModel';
import { Absorption, KineticParameters } from '../adapters/functionConverter';
import { Interactant } from '../adapters/interactionConverter';
import { FeatureDatum } from '../components/protein-data-views/UniProtKBFeaturesView';
import { Evidence } from './modelTypes';

export type FreeTextType =
  | 'DISRUPTION PHENOTYPE'
  | 'DOMAIN'
  | 'FUNCTION'
  | 'INDUCTION'
  | 'MISCELLANEOUS'
  | 'PATHWAY'
  | 'PTM'
  | 'SIMILARITY'
  | 'SUBUNIT'
  | 'TISSUE SPECIFICITY'
  | 'POLYMORPHISM'
  | 'ACTIVITY REGULATION'
  | 'BIOTECHNOLOGY'
  | 'CAUTION'
  | 'DEVELOPMENTAL STAGE';

export type CommentType =
  | FreeTextType
  | 'ALLERGEN'
  | 'ALTERNATIVE PRODUCTS'
  | 'BIOPHYSICOCHEMICAL PROPERTIES'
  | 'CATALYTIC ACTIVITY'
  | 'CAUTION'
  | 'COFACTOR'
  | 'DEVELOPMENTAL STAGE'
  | 'DISEASE'
  | 'INTERACTION'
  | 'MASS SPECTROMETRY'
  | 'PHARMACEUTICAL'
  | 'RNA EDITING'
  | 'SEQUENCE CAUTION'
  | 'SUBCELLULAR LOCATION'
  | 'TOXIC DOSE'
  | 'UNKNOWN'
  | 'WEB RESOURCE';

export type TextWithEvidence = {
  value: string;
  evidences?: Evidence[];
  id?: string;
};

export interface GenericComment<T extends CommentType = CommentType> {
  commentType: T;
  molecule?: string;
}

export interface FreeTextComment extends GenericComment<FreeTextType> {
  texts?: TextWithEvidence[];
}

export interface AbsorptionComment
  extends GenericComment<'BIOPHYSICOCHEMICAL PROPERTIES'> {
  absorption?: Absorption;
}

export interface KineticsComment
  extends GenericComment<'BIOPHYSICOCHEMICAL PROPERTIES'> {
  kineticParameters?: KineticParameters;
}

export interface pHDependenceComment extends KineticsComment {
  phDependence: {
    texts: TextWithEvidence[];
  };
}

export interface RedoxPotentialComment
  extends GenericComment<'BIOPHYSICOCHEMICAL PROPERTIES'> {
  redoxPotential: {
    texts: TextWithEvidence[];
  };
}

export interface TemperatureDependenceComment
  extends GenericComment<'BIOPHYSICOCHEMICAL PROPERTIES'> {
  temperatureDependence: {
    texts: TextWithEvidence[];
  };
}

export type PhysiologicalReactionDirection = 'left-to-right' | 'right-to-left';

export type PhysiologicalReaction = {
  directionType: PhysiologicalReactionDirection;
  reactionCrossReference: { database: string; id: string };
  evidences: Evidence[];
};

export interface CatalyticActivityComment
  extends GenericComment<'CATALYTIC ACTIVITY'> {
  reaction?: {
    name: string;
    reactionCrossReferences?: { database: string; id: string }[];
    ecNumber?: string;
    evidences?: Evidence[];
  };
  physiologicalReactions?: PhysiologicalReaction[];
}

export type DiseaseType = {
  diseaseId?: string;
  diseaseAccession?: string;
  acronym?: string;
  description?: string;
  diseaseCrossReference?: Xref;
  evidences?: Evidence[];
};

export interface DiseaseComment extends GenericComment<'DISEASE'> {
  disease?: DiseaseType;
  note?: { texts?: TextWithEvidence[] };
}

export enum InteractionType {
  SELF = 'SELF',
  XENO = 'XENO',
  BINARY = 'BINARY',
  // Are there others?
}

export type Interaction = {
  numberOfExperiments: number;
  type?: InteractionType; // Note: this has been removed. Use organismDiffer to know if Xeno.
  interactantOne: Interactant;
  interactantTwo: Interactant;
  organismDiffer: boolean;
};

export interface InteractionComment extends GenericComment<'INTERACTION'> {
  interactions: Interaction[];
}

export type Isoform = {
  name: TextWithEvidence;
  isoformSequenceStatus: string;
  isoformIds: string[];
  synonyms?: TextWithEvidence[];
  note?: { texts: TextWithEvidence[] };
  sequenceIds?: string[];
  varSeqs?: FeatureDatum[];
};

export interface AlternativeProductsComment
  extends GenericComment<'ALTERNATIVE PRODUCTS'> {
  isoforms: Isoform[];
  note?: { texts: TextWithEvidence[] };
  events: string[];
}

export interface SequenceCautionComment
  extends GenericComment<'SEQUENCE CAUTION'> {
  sequenceCautionType: string;
  sequence: string;
  note?: string;
  evidences?: Evidence[];
}

export interface MassSpectrometryComment
  extends GenericComment<'MASS SPECTROMETRY'> {
  method?: string;
  note?: string;
  molWeight: number;
  molWeightError: number;
  evidences: Evidence[];
}

export interface RNAEditingComment extends GenericComment<'RNA EDITING'> {
  locationType?: string;
  positions: { position: number; evidences: Evidence[] }[];
  note?: { texts: TextWithEvidence[] };
}

export interface SubcellularLocationComment
  extends GenericComment<'SUBCELLULAR LOCATION'> {
  note?: { texts: TextWithEvidence[] };
  molecule?: string;
  subcellularLocations?: {
    location: TextWithEvidence;
    topology?: TextWithEvidence;
    orientation?: TextWithEvidence;
  }[];
}

export interface WebResourceComment extends GenericComment<'WEB RESOURCE'> {
  note?: string;
  resourceName: string;
  resourceUrl: string;
  ftp?: boolean;
}

export interface CofactorComment extends GenericComment<'COFACTOR'> {
  cofactors?: {
    name: string;
    evidences?: Evidence[];
    cofactorCrossReference?: Xref;
  }[];
  note?: {
    texts: TextWithEvidence[];
  };
}

type Comment =
  | FreeTextComment
  | CatalyticActivityComment
  | DiseaseComment
  | InteractionComment
  | AlternativeProductsComment
  | SequenceCautionComment
  | SubcellularLocationComment
  | MassSpectrometryComment
  | RNAEditingComment
  | AbsorptionComment
  | KineticsComment
  | CofactorComment
  | pHDependenceComment
  | RedoxPotentialComment
  | TemperatureDependenceComment
  | SubcellularLocationComment
  | WebResourceComment;

export default Comment;
