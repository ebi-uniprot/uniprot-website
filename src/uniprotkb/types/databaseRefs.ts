export enum DatabaseCategory {
  CHEMISTRY = 'CHEMISTRY',
  DOMAIN = 'FMD',
  EXPRESSION = 'GEP',
  FAMILY = 'PFAM',
  GEL = '2DG',
  GENOME = 'GMA',
  INTERACTION = 'PPI',
  MISCELLANEOUS = 'MISC',
  ORGANISM = 'ORG',
  PATHWAY = 'EAP',
  PHYLOGENOMIC = 'PLG',
  GENETIC_VARIATION = 'GVD',
  GENE_ONTOLOGY = 'OTG',
  PROTEOMIC = 'PROTEOMIC',
  PROTEOMES = 'PRM',
  PROTOCOL = 'PAM',
  PTM = 'PTM',
  SEQUENCE = 'SEQ',
  STRUCTURE = '3DS',
}

export type AttributesItem = {
  name: string;
  xmlTag: string;
  uriLink?: string;
};

export type DatabaseInfoPoint = {
  name: string;
  displayName: string;
  category: string; // TODO: set as category type
  uriLink?: string;
  attributes?: AttributesItem[];
  implicit?: boolean;
  linkedReason?: string;
  idMappingName?: string;
};

export type DatabaseInfo = DatabaseInfoPoint[];
