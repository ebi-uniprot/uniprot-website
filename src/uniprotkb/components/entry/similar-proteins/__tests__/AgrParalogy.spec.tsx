import { AgrParalogsResult } from '../../../../types/agrParalogs';
import { columns, getRowId } from '../AgrParalogy';
import getHomologyMethodColumnConfig from '../getHomologyMethodColumnConfig';

jest.mock('../getHomologyMethodColumnConfig', () =>
  jest.fn((index, method, tooltip, commonTooltip, accessor) => ({
    id: `col-${index}`,
    method,
    tooltip,
    commonTooltip,
    accessor: accessor({
      geneToGeneParalogy: { objectGene: { foo: 'bar' } },
    }),
  }))
);

describe('getRowId()', () => {
  it('returns the primaryExternalId of the objectGene', () => {
    const result = {
      geneToGeneParalogy: {
        objectGene: { primaryExternalId: 'XYZ:1234' },
      },
    } as AgrParalogsResult;
    expect(getRowId(result)).toBe('XYZ:1234');
  });
});

describe('columns array', () => {
  const PARALOGY_METHOD_COUNT = 10;
  it('has the right total length', () => {
    expect(columns).toHaveLength(5 + PARALOGY_METHOD_COUNT + 1);
  });

  it('starts with the 5 static IDs', () => {
    expect(columns.slice(0, 5).map((c) => c.id)).toEqual([
      'gene-symbol',
      'rank',
      'length',
      'similarity',
      'identity',
    ]);
  });

  it('ends with "method-match-count"', () => {
    expect(columns[columns.length - 1].id).toBe('method-match-count');
  });

  it('calls getHomologyMethodColumnConfig for each method', () => {
    const calls = (getHomologyMethodColumnConfig as jest.Mock).mock.calls;
    expect(calls).toHaveLength(PARALOGY_METHOD_COUNT);
    const [index, method, tooltip, commonTooltip, accessor] = calls[0];
    expect(typeof index).toBe('number');
    expect(typeof method).toBe('string');
    expect(typeof tooltip).toBe('string');
    expect(commonTooltip).toMatch(/Result of paralogy-inference/);
    expect(
      accessor({ geneToGeneParalogy: { objectGene: { foo: 'bar' } } })
    ).toEqual({ objectGene: { foo: 'bar' } });
  });
});
