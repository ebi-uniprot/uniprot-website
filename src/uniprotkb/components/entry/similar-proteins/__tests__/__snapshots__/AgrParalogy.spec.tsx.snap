// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgrOrthology should render 1`] = `
<DocumentFragment>
  <div
    class="agr-link"
  >
    The data in this table is sourced from the Alliance of Genome Resources.
    <br />
    <a
      class="external-link"
      href="https://www.alliancegenome.org/gene/HGNC:620#paralogy"
      rel="noopener"
      target="_blank"
    >
      View source data
      <test-file-stub
        data-testid="external-link-icon"
        width="12.5"
      />
    </a>
     | 
    <a
      class="external-link"
      href="//www.alliancegenome.org/help"
      rel="noopener"
      target="_blank"
    >
      View source help
      <test-file-stub
        data-testid="external-link-icon"
        width="12.5"
      />
    </a>
  </div>
  <table
    class="table agr-homology-table agr-paralogy-table"
  >
    <thead
      class=""
    >
      <tr
        class="row"
      >
        <th />
        <th>
          Gene Symbol
        </th>
        <th>
          Rank
        </th>
        <th>
          Length
        </th>
        <th>
          Similarity
        </th>
        <th>
          Identity
        </th>
        <th>
          <div
            class="methods-label-container"
          >
            <span
              class="methods-label"
            >
              Method
            </span>
            <div
              class="method-label"
            >
              <span>
                Ensembl Compara
              </span>
            </div>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              HGNC
            </span>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              InParanoid
            </span>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              OMA
            </span>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              OrthoFinder
            </span>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              OrthoInspector
            </span>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              PANTHER
            </span>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              PhylomeDB
            </span>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              SonicParanoid
            </span>
          </div>
        </th>
        <th>
          <div
            class="method-label"
          >
            <span>
              SGD
            </span>
          </div>
        </th>
        <th>
          <span
            class="method-match-count-label"
          >
            <span>
              Match count
            </span>
          </span>
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class="row"
        data-id="HGNC:598"
      >
        <td />
        <td>
          <a
            href="/uniprotkb?query=%28taxonomy_id%3A9606%29+AND+%28xref%3Ahgnc-598%29"
          >
            APLP2
          </a>
        </td>
        <td>
          1
        </td>
        <td>
          810
        </td>
        <td>
          66
        </td>
        <td>
          50
        </td>
        <td>
          <span
            class="method-render"
            title="Match by Ensembl Compara method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Comparison not available for HGNC method"
          >
            -
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by InParanoid method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="No match by OMA method"
          >
            ○
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by OrthoFinder method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by OrthoInspector method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by PANTHER method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by PhylomeDB method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by SonicParanoid method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Comparison not available for SGD method"
          >
            -
          </span>
        </td>
        <td>
          <span
            class="method-match-count-render"
            title="7 matches from 8 checked methods (88%)"
          >
            7 of 8
          </span>
        </td>
      </tr>
      <tr
        class="row odd"
        data-id="HGNC:597"
      >
        <td />
        <td>
          <a
            href="/uniprotkb?query=%28taxonomy_id%3A9606%29+AND+%28xref%3Ahgnc-597%29"
          >
            APLP1
          </a>
        </td>
        <td>
          2
        </td>
        <td>
          776
        </td>
        <td>
          51
        </td>
        <td>
          36
        </td>
        <td>
          <span
            class="method-render"
            title="Match by Ensembl Compara method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Comparison not available for HGNC method"
          >
            -
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="No match by InParanoid method"
          >
            ○
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="No match by OMA method"
          >
            ○
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by OrthoFinder method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by OrthoInspector method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by PANTHER method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by PhylomeDB method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Match by SonicParanoid method"
          >
            ●
          </span>
        </td>
        <td>
          <span
            class="method-render"
            title="Comparison not available for SGD method"
          >
            -
          </span>
        </td>
        <td>
          <span
            class="method-match-count-render"
            title="6 matches from 8 checked methods (75%)"
          >
            6 of 8
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
